[TG-3374] Fix generic_type_index

romainbrenguier opened the pull request:
This function needs to look at the type_variables comment to known which
argument corresponds to the given type.
The bug was causing problems in the case of interfaces used with two different types in the generic arguments. The added test is an example of this.

romainbrenguier committed:
Fix generic_type_index

This function needs to look at the type_variables comment to known which
argument correspondsd to the given type.

romainbrenguier committed:
Test for abstract class with two generic arguments

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -611,10 +611,12 @@ class java_generic_symbol_typet : public symbol_typet
-      std::find(generic_types().begin(), generic_types().end(), type);
-    if(type_pointer != generic_types().end())
-      return type_pointer - generic_types().begin();
+    const auto type_variables = type.find(ID_type_variables);
```
`const auto &type_variables`

majakusber replied:
I ran the example in the test to understand better where the problem was. What happens here basically is that if in this function we are trying to find an index of a generic parameter (rather then a concrete type), the function does not work well. This is because the name of the parameter is a comment in the irep and thus does not get compared. The name of the parameter is in ID_type_variables, that's why you need to pull it out here. In order to make it more clear is happening, could you please use `std::find_if` and make use of the functions `is_java_generic_parameter`, `to_java_generic_parameter().get_name()`?

tautschnig commented on src/java_bytecode/java_types.h:
```diff
@@ -611,10 +611,12 @@ class java_generic_symbol_typet : public symbol_typet
+    {
+      if(type_variables == generic_types()[i].find(ID_type_variables))
+        return i;
+    }
```
This approach will do a lot of calls to `generic_types()`. At bare minimum create a temporary, but you might also do:
````c++
std::size_t index = 0;
for(const auto &generic_type : generic_types())
{
  if(type_variables == generic_type.find(ID_type_variables))
    return index;

  ++index;
}
````

majakusber reviewed (changes requested):
Also discussed offline, the implementation could be made more understandable by using existing functions.

thk123 commented on regression/jbmc-cover/generics/AbstractTest.java:
```diff
@@ -0,0 +1,23 @@
+        AbstractImpl<Dummy, Dummy> dummy = new AbstractImpl<>();
+        ClassB b = arg.get();
+        int i = b.getId();
+	return i;
```
Nit: dodgy formatting

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -609,12 +609,18 @@ class java_generic_symbol_typet : public symbol_typet
   /// \return The index of the type in the vector of generic types.
-  optionalt<size_t> generic_type_index(const reference_typet &type) const
+  optionalt<size_t>
+  generic_type_index(const java_generic_parametert &type) const
```
If you were to write a trivial unit test for this function that would be :ok_hand: to easily see why it was broken before but not now.

thk123 reviewed (approved):
I'd really like a unit test - the change seems correct but I can't follow what was wrong before and why the attached ticket failed it.

romainbrenguier committed:
Unit test java_type_from_string and fix doc

Unit-testing this function revealed some mistakes in the examples given
in the documentation.

romainbrenguier committed:
Move java_generic_symbol_typet definitions to cpp

romainbrenguier committed:
Unit test for java_generic_symbol_type

romainbrenguier committed:
Unit test for generic_type_index

romainbrenguier commented:
@thk123 I added some unit tests for this and for two other functions on which my test depended

thk123 commented on src/java_bytecode/java_types.cpp:
```diff
@@ -833,3 +835,34 @@ void get_dependencies_from_generic_parameters(
+      return i;
+  }
+  return {};
+}
```
Nit: missing new line

thk123 commented on src/java_bytecode/java_types.h:
```diff
@@ -579,21 +579,17 @@ class java_generic_symbol_typet : public symbol_typet
 public:
   typedef std::vector<reference_typet> generic_typest;
 
+  /// Construct a generic symbol type from base_ref and class_name_prefix which
```
Docs live with the implementation

thk123 commented on unit/java_bytecode/java_types/generic_type_index.cpp:
```diff
@@ -0,0 +1,33 @@
+
+SCENARIO("generic_type_index", "[core][java_types]")
+{
+  // Arrange
```
More "in style" to use the Catch keywords `GIVEN` `WHEN` and `THEN` instead of arrange/act/assert

thk123 commented on unit/java_bytecode/java_types/generic_type_index.cpp:
```diff
@@ -0,0 +1,33 @@
+  // Arrange
+  const auto symbol_type = symbol_typet("MyType");
+  const auto generic_symbol_type = java_generic_symbol_typet(
+    symbol_type, "LGenericClass<TX;TY;>;", "PrefixClassName");
```
Suggested additions: Non-single letter generic parameter name, name corresponding to a real class

thk123 commented on unit/java_bytecode/java_types/generic_type_index.cpp:
```diff
@@ -0,0 +1,33 @@
+  REQUIRE(indexX.value() == 0);
+  REQUIRE(indexY.has_value());
+  REQUIRE(indexY.value() == 1);
+  REQUIRE(!indexZ.has_value());
```
Nit: Prefer `REQUIRE_FALSE` as harder to miss than `!`

thk123 reviewed (approved):
Thanks for the tests they look really good. I've got a couple of stylistic suggestions but nothing blocking

thk123 commented on unit/java_bytecode/java_types/java_type_from_string.cpp:
```diff
@@ -0,0 +1,48 @@
+{
+  WHEN("Ljava/lang/Integer;")
+  {
+    const auto generic_symbol_type =
```
`generic_symbol_type` isn't a great variable name. If you could also put a TODO in these tests to indicate they aren't very comprehensive :stuck_out_tongue:

majakusber commented on src/java_bytecode/java_types.h:
```diff
@@ -609,12 +609,18 @@ class java_generic_symbol_typet : public symbol_typet
-      std::find(generic_types().begin(), generic_types().end(), type);
-    if(type_pointer != generic_types().end())
-      return type_pointer - generic_types().begin();
+    const auto &type_variable = type.type_variable();
```
Instead of `type_variable` use `get_name()` to get the unique name of the parameter.

majakusber commented on src/java_bytecode/java_types.h:
```diff
@@ -609,12 +609,18 @@ class java_generic_symbol_typet : public symbol_typet
+    {
+      if(
+        is_java_generic_parameter(generics[i]) &&
+        to_java_generic_parameter(generics[i]).type_variable() == type_variable)
```
Use `get_name()` again, see above.

majakusber commented on src/java_bytecode/java_types.h:
```diff
@@ -609,12 +609,18 @@ class java_generic_symbol_typet : public symbol_typet
   /// \return The index of the type in the vector of generic types.
-  optionalt<size_t> generic_type_index(const reference_typet &type) const
+  optionalt<size_t>
+  generic_type_index(const java_generic_parametert &type) const
```
Please update the documentation to reflect that only a generic parameter can be passed here. Also, wherever this function is called, make sure that it's only called on a parameter - for example by adding preconditions if it's not obvious.

romainbrenguier replied:
The type already ensures it is only called on a java_generic_parametert so I don't think a precondition is necessary

majakusber replied:
I meant a precondition before a call to this function, elsewhere in the code.

romainbrenguier replied:
If the function is called then the argument is a java_generic_parametert so there shouldn't be anything to check?

majakusber replied:
Yes, that's exactly what I meant - if the object that this function is called with is already typed as `java_generic_parametert` then there is nothing to check.

romainbrenguier committed:
Add java_types unit tests to Makefile

majakusber commented on src/java_bytecode/java_types.h:
```diff
@@ -610,19 +599,7 @@ class java_generic_symbol_typet : public symbol_typet
   /// \param type The type we are looking for.
   /// \return The index of the type in the vector of generic types.
   optionalt<size_t>
-  generic_type_index(const java_generic_parametert &type) const
```
Move the documentation too.

majakusber commented on src/java_bytecode/java_types.h:
```diff
@@ -579,6 +579,13 @@ class java_generic_symbol_typet : public symbol_typet
+  /// For instance `java_generic_symbol_typet(symbol_typet("Map"),
+  /// "Ljava/util/Map<TK;TV;>;", "java/util/Map")` will generate a symbol type
+  /// with identifier "Map", and two generic types with identifier
+  /// "java/util/Map::K" and "java/util/Map::V" respectively.
```
To reflect what this does a bit better I would say - Construct a generic symbol type by extending the symbol type \p type with generic types extracted from the reference \p base_ref. This assumes that the class with name \p class_name_prefix extends/implements the class in \p type.

For the example you give, I would change the third input to "java/util/HashMap" and then the generic types identifiers become "java.util.HashMap::K" and "java.util.HashMap::V".

majakusber replied:
Actually, there should be a precondition in the constructor that checks that the identifier of `type` and the identifier of the type created for `base_ref` are the same. But that is up to you if you want to add this here, I can do that in a separate PR.

majakusber commented on unit/java_bytecode/java_types/java_generic_symbol_type.cpp:
```diff
@@ -0,0 +1,31 @@
+  {
+    auto symbol_type = symbol_typet("MyType");
+    const auto generic_symbol_type = java_generic_symbol_typet(
+      symbol_type, "LGenericClass<TX;TY;>;", "PrefixClassName");
```
Given the missing precondition in the constructor, as mentioned above, replace this with `"LMyType<TX;TY;>;"`.

majakusber commented on unit/java_bytecode/java_types/generic_type_index.cpp:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  GIVEN("Generic type LGenericClass<Tkey;Tvalue;>; and"
+          " parameters key, value, x")
```
How is this example different from the previous one?

romainbrenguier replied:
It has non-single letter generic parameter names as suggested by @thk123

majakusber replied:
I see. Maybe you can combine them into one example though, with one single letter and one non-single letter case.

majakusber reviewed (changes requested):
Looks good but still few adjustments needed.

romainbrenguier committed:
Use get_name in generic_type_index

romainbrenguier committed:
Add invariant in java_generic_symbol_type

This also adapt unit tests that would make the invariant fail.

romainbrenguier commented:
@svorenova I made the changes, can you check again?

majakusber commented on src/java_bytecode/java_types.cpp:
```diff
@@ -854,6 +854,11 @@ java_generic_symbol_typet::java_generic_symbol_typet(
   const java_generic_typet gen_base_type = to_java_generic_type(base_type);
+  INVARIANT(
+    type.get_identifier() == to_symbol_type(gen_base_type.subtype()).get_identifier(),
+    std::string("identifier of ")+type.pretty()+"\n and identifier of type "+
```
Please use double quotes here as well instead of `std::string`.

majakusber commented on src/java_bytecode/java_types.cpp:
```diff
@@ -833,3 +835,51 @@ void get_dependencies_from_generic_parameters(
+/// For instance `java_generic_symbol_typet(symbol_typet("Map"),
+/// "Ljava/util/Map<TK;TV;>;", "java/util/Map")` will generate a symbol type
+/// with identifier "Map", and two generic types with identifier
+/// "java/util/Map::K" and "java/util/Map::V" respectively.
```
My previous comment on changing  this example is still outstanding: please use `java.util.HashMap` as the third argument and the parameter identifiers are then `java.util.HashMap::K` and `java.util.HashMap::V`. This then correctly corresponds to the case `HashMap<K,V> extends Map<K,V>`.

majakusber commented on unit/java_bytecode/java_types/generic_type_index.cpp:
```diff
@@ -0,0 +1,65 @@
+    }
+  }
+
+  GIVEN("Generic type Ljava/util/HashMap<TK;TV;>; and parameters K, V, T")
```
Sorry for being nit picking here but as I thought about the application of `java_generic_symbol_typet` I would suggest to change these examples a bit to reflect why that type even exists. It is to represent the generic base of a class. So you could change the examples to say that for example given `HashMap<K,V> extends Map<K,V>` the generic symbol for Map is created correctly - the generic symbol is then created with `java_generic_symbol_typet(symbol_typet("java::java.util.Map"), "Ljava/util/Map<TK;TV;>;", "java.util.HashMap")` (exactly as in the documentation for the constructor actually). Similarly for the first example - `PrefixClassName<X,value> extends GenericClass<X,value>` and the constructor call stays as it is.

majakusber reviewed (approved):
Looks good, only two minor suggestions to make the examples a bit more realistic. Please also make sure that you squash related commits before merging.
